src/platforms/hubspot/client.ts

```diff
@@ -0,0 +1,336 @@
+import { makeRequestFactory } from '@/sdk/client';
+import * as z from 'zod';
+import { API_VERSION, BASE_URL, HUBSPOT_MAX_PAGE_SIZE } from './constants';
+import {
+  HubspotCallCreate,
+  hubspotCallSchema,
+  HubspotCallUpdate,
+  HubspotCompanyCreate,
+  hubspotCompanySchema,
+  HubspotCompanyUpdate,
+  HubspotContactCreate,
+  hubspotContactListSchema,
+  listResponseHubspotContactListSchema,
+  hubspotContactSchema,
+  HubspotContactUpdate,
+  HubspotDealCreate,
+  hubspotDealSchema,
+  HubspotDealUpdate,
+  HubspotEmailCreate,
+  hubspotEmailSchema,
+  HubspotEmailUpdate,
+  HubspotMeetingCreate,
+  hubspotMeetingSchema,
+  HubspotMeetingUpdate,
+  HubspotModule,
+  HubspotNoteCreate,
+  hubspotNoteSchema,
+  HubspotNoteUpdate,
+  hubspotOwnerSchema,
+  HubspotTaskCreate,
+  hubspotTaskSchema,
+  HubspotTaskUpdate,
+  listResponseSchema,
+  listResponseHubspotContactListContactsSchema,
+  ListObjectInput,
+  FindObjectInput,
+  hubspotPropertySchema,
+  HubspotCustomPropertyCreate,
+  HubspotAssociationCreate,
+  hubspotAssociationLabelOutputSchema,
+  HubspotAssociationLabelInput,
+  contactProperties,
+  companyProperties,
+  dealProperties,
+  noteProperties,
+  taskProperties,
+  meetingProperties,
+  emailProperties,
+  callProperties,
+  ListOutput,
+  HubspotOwner,
+  HubspotContact,
+  HubspotCompany,
+  HubspotDeal,
+  HubspotNote,
+  HubspotTask,
+  HubspotMeeting,
+  HubspotEmail,
+  HubspotCall,
+  BatchReadObjectInput,
+} from './schemas';
+import { omit, shake } from 'radash';
+import { Auth, ClientResult } from '@/sdk';
+
+const request = makeRequestFactory(async (auth, options) => {
+  return {
+    ...options,
+    url: `${BASE_URL}${options.url}`,
+    headers: {
+      ...options.headers,
+      Authorization: `Bearer ${await auth.getToken()}`,
+    },
+  };
+});
+
+type requestFunctionType<I, O> = (
+  auth: Auth,
+  input: I,
+) => Promise<ClientResult<O>>;
+
+const makeClient = () => {
+  const findObject = <TOutput>(
+    module: HubspotModule | `objects/${HubspotModule}`,
+    schema: z.ZodSchema,
+    properties?: string[],
+  ): requestFunctionType<FindObjectInput, TOutput> =>
+    request(({ id, associations }: FindObjectInput) => ({
+      url: `/crm/${API_VERSION}/${module}/${id}`,
+      method: 'GET',
+      query: shake({
+        properties: properties?.join(','),
+        associations: associations?.join(','),
+      }),
+      schema,
+    }));
```

Comment:
this approach is more DRY but it's also a little hard to read, are all the types working fine?

Reply:
Yeah, the generics make the typing work out